DataDog/chaos-controller
# Add retry logic to dns resolution

### What does this PR do?

Adds logic for retrying dns resolution in case of common errors like timeouts.

### Motivation

Resolution should be given another shot when common errors occur because they are usually intermittent and not consistent.

### Testing Guidelines

I used the following code to test retry logic:
```
	///THIS IS FOR DEBUGGING PURPOSES, THIS SHOULD BE REMOVED!!!
	real_address := dnsConfig.Servers[0]
	dnsConfig.Servers[0] = "192.0.2.10"
	intentional_fails := 2
	/////////////////////////////////////////////////////////////
	response, _, err := dnsClient.Exchange(&dnsMessage, dnsConfig.Servers[0]+":53")
	for err != nil && retry < 4 {
		log.Info(fmt.Sprintf("Failed Resolution of Host %s on attempt %d: Failure: %s...", host, retry+1, err))
		time.Sleep(5 * time.Second)
		if retry == 3 {
			log.Info("This was the last retry...")
			return nil, fmt.Errorf("can't resolve the given hostname %s: %w", host, err)
		}
		///THIS IS FOR DEBUGGING PURPOSES, THIS SHOULD BE REMOVED!!!
		intentional_fails -= 1
		if intentional_fails == 0 {
			dnsConfig.Servers[0] = real_address
		}
		/////////////////////////////////////////////////////////////
		retry += 1
		response, _, err = dnsClient.Exchange(&dnsMessage, dnsConfig.Servers[0]+":53")
	}
```
The first two attempts, it attempts to use a nameserver that does not exists, so it will timeout the first two times. Then it will use the correct nameserver and should succeed on the third try. Here are the results of this testing:
```
{"level":"info","ts":1596128172.954012,"caller":"injector/network_disruption.go:54","msg":"injecting network disruption"}
{"level":"info","ts":1596128172.9569514,"caller":"injector/network_disruption.go:75","msg":"adding network disruptions","drop":10,"corrupt":5}
{"level":"info","ts":1596128172.957138,"caller":"injector/network_config.go:122","msg":"auto-detecting interfaces to apply disruption to..."}
{"level":"info","ts":1596128172.95718,"caller":"injector/network_config.go:65","msg":"auto-detecting interfaces to apply disruption to..."}
{"level":"info","msg":"Failed Resolution of Host datadoghq.com on attempt 1: Failure: read udp 10.1.0.25:38186->192.0.2.10:53: i/o timeout..."}
{"level":"info","msg":"Failed Resolution of Host datadoghq.com on attempt 2: Failure: read udp 10.1.0.25:43352->192.0.2.10:53: i/o timeout..."}
{"level":"info","ts":1596128187.0695703,"caller":"injector/network_config.go:84","msg":"IP 13.226.217.93/32 belongs to interface eth0"}
{"level":"info","ts":1596128187.0696688,"caller":"injector/network_config.go:84","msg":"IP 13.226.217.127/32 belongs to interface eth0"}
{"level":"info","ts":1596128187.0697675,"caller":"injector/network_config.go:84","msg":"IP 13.226.217.6/32 belongs to interface eth0"}
{"level":"info","ts":1596128187.0698574,"caller":"injector/network_config.go:84","msg":"IP 13.226.217.60/32 belongs to interface eth0"}
{"level":"info","ts":1596128187.0698943,"caller":"injector/network_config.go:149","msg":"setting tx qlen for interface eth0"}
{"level":"info","ts":1596128187.0779135,"caller":"injector/network_config.go:210","msg":"clearing tx qlen for interface eth0"}
{"level":"info","ts":1596128187.0782294,"caller":"injector/network_disruption.go:92","msg":"operations applied successfully"}
{"level":"info","ts":1596128187.0782535,"caller":"injector/main.go:44","msg":"closing metrics sink client before exiting","sink":"noop"}
```
As you can see it correctly fails twice in the beginning and then completes successfully when the real nameserver from `/etc/resolv.conf` is used.
